sensor/common/registry/lazy_tls_image_registry.go

```diff
@@ -84,17 +84,24 @@ func (l *lazyTLSCheckRegistry) Metadata(image *storage.Image) (*storage.ImageMet
 	// Attempt initialization since Metadata interacts with the registry.
 	l.lazyInit()
 
-	// initError and registry are modified while the
+	err := l.getInitError()
+	if err != nil {
+		return nil, err
+	}
+
+	// At this point lazy init has successfully completed.
+	return l.registry.Metadata(image)
+}
+
+func (l *lazyTLSCheckRegistry) getInitError() error {
```

Comment:
Extracting that to a separate function is totally fine, but FYI we also have constructs like https://github.com/stackrox/stackrox/blob/4430f5c979ef4dbf84f9ba97e8192d2d4bf57b6e/pkg/concurrency/lock.go#L65 (there are also variants for reading only) that are perfectly suited for limiting the lock scope.

Reply:
+1